include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,311 @@
+/**
+ * @file nsga2_impl.hpp
+ * @author Sayan Goswami
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more Information.
+ */
+
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+#define ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+
+#include "nsga2.hpp"
+#include <set>
+#include <vector>
+#include <map>
+
+namespace ens {
+
+inline NSGA2::NSGA2(const size_t populationSize,
+                    const size_t maxGenerations,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const double epsilon) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    epsilon(epsilon)
+{ /* Nothing to do here. */ }
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType,
+         typename... CallbackTypes>
+std::vector<MatType> NSGA2::Optimize(MultiobjectiveFunctionType& objectives,
+                        MatType& iterate,
+                        CallbackTypes&&... callbacks)
+{
+  std::vector<arma::vec> calculatedObjectives;
```

Comment:
Do you think we could instead use an `arma::mat` since the size never changes?

Reply:
The default template parameter is `arma::mat` during instantiation (from `SchafferFunctionN1`). Should we explicitly state it here too?

Reply:
Yes I suppose that would make sense

Reply:
Also, it's not guaranteed that the return type is `double`, so we should use something like: `typedef typename MatType::elem_type ElemType;` to be generic.

Reply:
On it.